src/App/Fossa/VSI/DynLinked/Internal/Lookup/APK.hs

```diff
@@ -0,0 +1,163 @@
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE TupleSections #-}
+
+module App.Fossa.VSI.DynLinked.Internal.Lookup.APK (
+  lookupDependencies,
+  constructLookupTables,
+  tryLookup,
+  SyftData (..),
+  SyftArtifact (..),
+  SyftArtifactMetadata (..),
+  SyftArtifactMetadataFile (..),
+  SyftLookupTable (..),
+) where
+
+import App.Fossa.VSI.DynLinked.Types (DynamicDependency (..), LinuxPackageManager (..), LinuxPackageMetadata (..), ResolvedLinuxPackage (..))
+import App.Fossa.VSI.DynLinked.Util (runningLinux)
+import Control.Algebra (Has)
+import Control.Effect.Diagnostics (Diagnostics, fatalText)
+import Control.Monad (join)
+import Control.Monad.Trans.Maybe (MaybeT (MaybeT), runMaybeT)
+import Data.Aeson (FromJSON, Result (Error, Success), ToJSON, Value, fromJSON, object, parseJSON, toJSON, withObject, (.:), (.=))
+import Data.Either (partitionEithers)
+import Data.Map (Map)
+import Data.Map qualified as Map
+import Data.String.Conversion (toText)
+import Data.Text (Text)
+import Effect.Exec (AllowErr (Never), Command (..), Exec, execJson)
+import Path (Abs, Dir, File, Path)
+
+-- | The idea here is that we look up what paths we can with apt and turn them into @DynamicDependency@.
+-- We then hand back leftovers and lookup results for the next resolution function.
+lookupDependencies :: (Has Diagnostics sig m, Has Exec sig m) => Path Abs Dir -> [Path Abs File] -> m ([Path Abs File], [DynamicDependency])
+lookupDependencies _ files | not runningLinux = pure (files, [])
+lookupDependencies root files = do
+  syft <- runSyft root
+  table <- constructLookupTables syft
+  partitionEithers <$> traverse (tryLookup table) files
+
+tryLookup :: (Has Diagnostics sig m) => SyftLookupTable -> Path Abs File -> m (Either (Path Abs File) DynamicDependency)
+tryLookup SyftLookupTable{..} file = fmap (maybeToRight file) . runMaybeT $ do
+  index <- MaybeT . pure $ Map.lookup file pathToIndex
+  meta <- MaybeT . pure $ Map.lookup index indexToMeta
+  pure . DynamicDependency file . Just $ ResolvedLinuxPackage LinuxPackageManagerRPM meta
```

Comment:
A few suggestions here:

- tryLookup doesn't actually use diagnostics, and can be rewritten to not use `MaybeT` as well.  Without those monads, this doesn't need to return `m Foo` at all.
- `maybeToRight` can be inlined as `maybe (Left file) Right`

You may also want to look at map composition: https://hackage.haskell.org/package/containers-0.6.5.1/docs/Data-Map-Internal.html#g:12

```suggestion
  pure . partitionEithers $ map (tryLookup table) files

tryLookup :: SyftLookupTable -> Path Abs File -> Either (Path Abs File) DynamicDependency
tryLookup SyftLookupTable{..} file = maybe (Left file) Right dyn
  where
    index = Map.lookup file pathToIndex
    meta = index >>= (`Map.lookup` indexToMeta)
    dyn :: Maybe DynamicDependency
    dyn = DynamicDependency file . Just . ResolvedLinuxPackage LinuxPackageManagerRPM <$> meta
```

Reply:
Aternatively, if you drop the `m` from `Diagnostics`, you can use the do notation for the `Maybe` monad, instead of its monad transformer:
```suggestion
  pure . partitionEithers $ map (tryLookup table) files
tryLookup :: SyftLookupTable -> Path Abs File -> Either (Path Abs File) DynamicDependency
tryLookup SyftLookupTable{..} file = maybe (Left file) Right $ do
  index <- Map.lookup file pathToIndex
  meta <- Map.lookup index indexToMeta
  Just . DynamicDependency file . Just $ ResolvedLinuxPackage LinuxPackageManagerRPM meta
```